adhocteam/Head-Start-TTADP
# Add file scanning

**Description of change**
Add support for scanning files using worker instance.


**How to test**
***Test happy path***
1. Pull down changes and start server. 
2. Wait for everything to come up (including clamav which takes a while since it updates the db the first time you run it)
3. Fill in a grantee and the go to topics and resources.
4. Upload a file this isn't plaintext, such as a docx or pdf
5. refresh page and you should see either scanning queued or approved.
6. If you see scanning queued refresh the page again and you should see approved.
7. Watch the logs and you will see a message like `worker_1       | {"message":"job 3e934f98-47b4-41d0-9e53-3a40cc565688.pdf completed with status 200 and result { Status: \"OK\", Description: \"\" }","level":"info","timestamp":"2021-02-17T18:27:23.442Z"`
8. `yarn docker:stop` and `yarn docker:reset`

***Test retry***
Same as above except don't wait for clamav to come all the way up. When the web ui is ready, you can start uploading.
Watch the logs, you will see the jobs fail and retry up to 5 times with an exponential backoff.
While this is happening if you refresh the web ui you should see a "scanning queued" status.
After you see a success message like above, refreshing the web ui should show a status of approved.

***Test infected file***
Same as the happy path, except use file [virus-test.zip](https://github.com/HHS/Head-Start-TTADP/files/5997753/virus-test.zip) (note, this isn't a virus, just the clamav test file zipped so it will work with this app).
Refreshing the page will give you a status of rejected and you should see a message in the logs that says `{"message":"job 49bcf537-4932-4b1b-9cbc-790cad8b23cc.zip completed with status 406 and result { Status: \"FOUND\", Description: \"Eicar-Signature\" }","level":"info","timestamp":"2021-02-17T18:32:14.287Z"}`

**Issue(s)**
* https://github.com/HHS/Head-Start-TTADP/issues/204

**Checklist**

    - API methods
    - Boundary and Data Flow Diagrams
    - [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions with the [Nygard template](https://github.com/joelparkerhenderson/architecture_decision_record/blob/master/adr_template_by_michael_nygard.md)
    - OSCAL templates completed when security controls are implemented or modified
